### PR TITLE
Use crm_attribute --list-options

### PR DESCRIPTION
**Pull request by liangxin1300**
#### Changes include
- Dev: ra: Get cluster option metadata from crm_attribute command
    
    pacemaker enables crm_attribute --list-options since 2.1.8 version,
    which is preferable over the "pacemaker-schedulerd",
    "pacemaker-controld", "pacemaker-based", or even older
    "pengine", "crmd", "cib".

- Dev: ra: Parsing select values in cluster option metadata
- Dev: ra: Show related info when "advanced", "generated" and "deprecated" detected in metadata
- Dev: doc: Drop the options for the old daemon names, pengine, crmd, cib and stonithd in the help/man

#### Show allowed values
```
crm(live/15sp5-1)configure# property no-quorum-policy=
no-quorum-policy (select, [stop]): 
    What to do when the cluster does not have quorum
    Allowed values: stop, freeze, ignore, demote, suicide
```

#### Show advanced option
```
crm(live/15sp5-1)configure# property election-timeout= 
election-timeout (time, [2min]): 
    *** Advanced Use Only ***
    Declare an election failed if it is not decided within this much time. If you need to adjust this value, it probably indicates the presence of a bug.
```
Also for the fence metadata
```
crm(live/15sp5-1)configure# primitive id=ddd stonith:fence_eps params pcmk_monitor_retries= 
pcmk_monitor_retries (integer, [2]): The maximum number of times to try the 'monitor' command within the timeout period
    *** Advanced Use Only ***
    Some devices do not support multiple connections. Operations may "fail" if the device is busy with another task. In that case, Pacemaker will automatically retry the operation if there is time remaining. Use this option to alter the number of times Pacemaker tries a 'monitor' action before giving up.
```
#### Show auto generated option
```
crm(live/15sp5-1)configure# property have-watchdog=
have-watchdog (boolean, [false]): Whether watchdog integration is enabled
    *** Automatically generated by pacemaker ***
    This is set automatically by the cluster according to whether SBD is detected to be in use. User-configured values are ignored. The value `true` is meaningful if diskless SBD is used and `stonith-watchdog-timeout` is nonzero. In that case, if fencing is required, watchdog-based self-fencing will be performed via SBD without requiring a fencing resource explicitly configured.
```
#### Show deprecated option
```
crm(live/15sp5-1)configure# property remove-after-stop= 
remove-after-stop (boolean, [false]): Whether to remove stopped resources from the executor
    *** Deprecated ***
    Values other than default are poorly tested and potentially dangerous.
```

#### Adjust help info at `crm ra info`
```
# crm ra help info
Show meta data for a RA

Show the meta-data of a resource agent type. This is where users
can find information on how to use a resource agent. To get cluster
option meta-data, use the cluster type. To get pacemaker-fenced
meta-data, use the pacemaker-fenced type.

Usage:

info [<class>:[<provider>:]]<type>
info <type> <class> [<provider>] (obsolete)

Example:

info apache
info ocf:pacemaker:Dummy
info stonith:ipmilan
info cluster
info pacemaker-fenced
```

### CODE REVIEW

File: crmsh/ui_ra.py [hunk 1]
```diff
@@ -106,20 +106,22 @@ def do_list(self, context, class_, provider_=None):
     @command.completers(complete_class_provider_type)
```
Comment (zzhou1) [inline crmsh/ui_ra.py:106]: The following patch could provide auto-completion for `crm ra info <tab><tab>`

```
--- a/crmsh/ui_ra.py
+++ b/crmsh/ui_ra.py
@@ -23,9 +23,10 @@ def complete_class_provider_type(args):
     '''
     c = args[-1]
     classes = ra.ra_classes()
+    l_1st_completion = [l+":" for l in classes] + ["cluster", "pacemaker-fenced"]
     if not c:
         # First completion:  agent classes
-        return [l+":" for l in classes]
+        return l_1st_completion
 
     ret = set([])
     providers = ra.ra_providers_all('ocf')
@@ -62,7 +63,7 @@ def complete_class_provider_type(args):
         return list(ret)
 
     # First completion:  agent classes
-    return [l+":" for l in classes]
+    return l_1st_completion
 
 class RA(command.UI):
     '''

```

File: doc/crm.8.adoc [hunk 2]
```diff
@@ -1536,10 +1536,9 @@ classes
 ==== `info` (`meta`)
 
 Show the meta-data of a resource agent type. This is where users
-can find information on how to use a resource agent. It is also
-possible to get information from some programs: `pengine`,
-`crmd`, `cib`, and `stonithd`. Just specify the program name
-instead of an RA.
+can find information on how to use a resource agent. To get cluster
+option meta-data, use the `cluster` type. To get pacemaker-fenced
+meta-data, use the `pacemaker-fenced` type.
```
Comment (zzhou1) [inline doc/crm.8.adoc:1541]:
```suggestion
can find information on how to use a resource agent. It is also
possible to use the `pacemaker-fenced` type to get the corresponding
meta-data information.
```

File: doc/crm.8.adoc [hunk 3]
```diff
@@ -1551,7 +1550,8 @@ Example:
 info apache
 info ocf:pacemaker:Dummy
 info stonith:ipmilan
-info pengine
+info cluster
```
Comment (zzhou1) [inline doc/crm.8.adoc:1553]:
```suggestion
```

If you agree, we could delete and replace "crm ra info cluster" to "crm configure property info".

**Comment by nicholasyang2022** [comment]
The text output from `crm_attribute --list-options cluster` is formatted better. So we also improve the formatting in crmsh?

From `crm ra info cluster`:

```text
Pacemaker cluster options (cluster_option:None)

Also known as properties, these are options that affect behavior across the entire cluster. They are configured within cluster_property_set elements inside the crm_config subsection of the CIB configuration section.

Parameters (*: required, []: default):

dc-version (string): Pacemaker version on cluster node elected Designated Controller (DC)
    *** Automatically generated by pacemaker ***
    Includes a hash which identifies the exact revision the code was built from. Used for diagnostic purposes.
```

From `crm_attribute --list-options cluster`:

```text
Pacemaker cluster options

Also known as properties, these are options that affect behavior across the entire cluster. They are configured within cluster_property_set elements inside the crm_config subsection of the CIB configuration section.

  * dc-version: Pacemaker version on cluster node elected Designated Controller (DC)
    * Includes a hash which identifies the exact revision the code was built from. Used for diagnostic purposes.
    * Possible values (generated by Pacemaker): string (no default)
```

**Comment by zzhou1** [comment]
> The text output from crm_attribute --list-options cluster is formatted better. So we also improve the formatting in crmsh?

> * dc-version: Pacemaker version on cluster node elected Designated Controller (DC)
>     * Includes a hash which identifies the exact revision the code was built from. Used for diagnostic purposes.

I like the idea. If possible, with the reasonable implementation effort, crmsh can pre-format those long line text to multiple lines with the space indentation for the small terminal. That will be great. More readable for user experience.

File: crmsh/ra.py [hunk 4]
```diff
@@ -213,6 +213,9 @@ def get_cib_meta():
 
 @utils.memoize
 def get_properties_meta():
+    cluster_option_meta = utils.get_cluster_option_metadata()
+    if cluster_option_meta:
+        return RAInfo("cluster_option", None, meta_string=cluster_option_meta)
```
[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/ra.py]: Changed

[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/ra.py]: Changed

**Comment by liangxin1300** [comment]
> The text output from `crm_attribute --list-options cluster` is formatted better. So we also improve the formatting in crmsh?
> 
> From `crm ra info cluster`:
> 
> ```
> Pacemaker cluster options (cluster_option:None)
> 
> Also known as properties, these are options that affect behavior across the entire cluster. They are configured within cluster_property_set elements inside the crm_config subsection of the CIB configuration section.
> 
> Parameters (*: required, []: default):
> 
> dc-version (string): Pacemaker version on cluster node elected Designated Controller (DC)
>     *** Automatically generated by pacemaker ***
>     Includes a hash which identifies the exact revision the code was built from. Used for diagnostic purposes.
> ```
> 
> From `crm_attribute --list-options cluster`:
> 
> ```
> Pacemaker cluster options
> 
> Also known as properties, these are options that affect behavior across the entire cluster. They are configured within cluster_property_set elements inside the crm_config subsection of the CIB configuration section.
> 
>   * dc-version: Pacemaker version on cluster node elected Designated Controller (DC)
>     * Includes a hash which identifies the exact revision the code was built from. Used for diagnostic purposes.
>     * Possible values (generated by Pacemaker): string (no default)
> ```

Chaned when calling `crm cluster ra info cluster`, use the text format directly

File: doc/crm.8.adoc (hunk 3 above)
Comment (liangxin1300) [reply doc/crm.8.adoc:1553]: I prefer not to hide it from help info, just don't show it from completion result

**Comment by zzhou1** [comment]
One more finding the following help text need be revised.

```
crm configure property --help
Set a cluster property

Set cluster configuration properties. To list the
available cluster configuration properties, use the
ra info command with pengine, crmd,
cib and stonithd as arguments.
```

**Comment by liangxin1300** [comment]
> Few more findings:
> 
> 1. Can we provide bash completion for `crm configure property <TAB><TAB>`? Just like the interactive mode which is very good.
> 2. Similar, to provide meta-data description completion, for example, `crm configure property stonith-timeout= <TAB><TAB>`

I will try to do above completion in a separate PR

> 3. the following help text need be revised accordingly.
> 
> ```
> crm configure property --help
> Set a cluster property
> 
> Set cluster configuration properties. To list the
> available cluster configuration properties, use the
> ra info command with pengine, crmd,
> cib and stonithd as arguments.
> ```

Already changed as https://github.com/ClusterLabs/crmsh/pull/1379/files#diff-9f089b36b9b7014682bc6ac27dccded1d47edc5fa14889679c64eae6dbc28649R3486
Does that still need to be hidden?

**Comment by zzhou1** [comment]
> 
> I will try to do above completion in a separate PR
> 

Sounds good.

> > 3. the following help text need be revised accordingly.
> > 

> Already changed as https://github.com/ClusterLabs/crmsh/pull/1379/files#diff-9f089b36b9b7014682bc6ac27dccded1d47edc5fa14889679c64eae6dbc28649R3486 Does that still need to be hidden?
> 
> Remove all?

Great! It makes sense to delay this to the next PR, and replace it with the new approach by using the bash completion <TAB><TAB>

**Review by zzhou1 — APPROVED** [review]

**Pull request merged**